python/dgl/nn/pytorch/explain/pgexplainer.py

```diff
@@ -0,0 +1,351 @@
+            Indicates whether the gate is being used during training or evaluation.
+
+        Returns
+        -------
+        Tensor
+            If training is set to True, the output is a tensor of probabilities that
+            represent the probability of activating the gate for each input element.
+            If training is set to False, the output is also a tensor of probabilities,
+            but they are determined solely by the log_alpha values, without adding any
+            random noise.
+        """
+        if training:
+            bias = self.sample_bias
+            random_noise = torch.rand(log_alpha.size()).to(log_alpha.device)
+            random_noise = bias + (1 - 2 * bias) * random_noise
+            gate_inputs = torch.log(random_noise) - torch.log(
+                1.0 - random_noise
+            )
+            gate_inputs = (gate_inputs + log_alpha) / beta
+            gate_inputs = torch.sigmoid(gate_inputs)
+        else:
+            gate_inputs = torch.sigmoid(log_alpha)
+
+        return gate_inputs
+
+    def train_step(self, graph, feat, tmp):
+        r"""Training the explanation network by gradient descent(GD)
+        using Adam optimizer
+
+        Parameters
+        ----------
+        graph : DGLGraph
+            A homogeneous graph.
+        feat : Tensor
+            The input feature of shape :math:`(N, D)`. :math:`N` is the
+            number of nodes, and :math:`D` is the feature size.
+        tmp : float
+            The temperature parameter fed to the sampling procedure.
+
+        Returns
+        -------
+        Tensor
+            A scalar tensor representing the loss.
+        """
+        self.model = self.model.to(graph.device)
+        self.elayers = self.elayers.to(graph.device)
+
+        pred = self.model(graph, feat).argmax(-1).data
+
+        prob, edge_mask = self.explain_graph(
+            graph, feat, tmp=tmp, training=True
+        )
+
+        self.edge_mask = edge_mask
+
+        loss_tmp = self.loss(prob.unsqueeze(dim=0), pred)
+        return loss_tmp
+
+    def explain_graph(self, graph, feat, tmp=1.0, training=False):
+        r"""Learn and return an edge mask that plays a crucial role to
+        explain the prediction made by the GNN for a graph. Also, return
+        the prediction made with the edges chosen based on the edge mask.
+
+        Parameters
+        ----------
+        graph : DGLGraph
+            A homogeneous graph.
+        feat : Tensor
+            The input feature of shape :math:`(N, D)`. :math:`N` is the
+            number of nodes, and :math:`D` is the feature size.
+        tmp : float
+            The temperature parameter fed to the sampling procedure.
+        training : bool
+            We indicate we want to train the explanation network if
+            set to True. If set to False, we indicate that we want
+            to find the edges of the graph that contribute most to
+            the graph explanations.
```

Comment:
Replace the paragraph with "Training the explanation network or simply generating explanations."

Reply:
fixed.